### Importing Packages

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

### Reading Dataset

In [2]:
## Reading Dataset
asgn_dt = pd.read_csv(r'C:\Users\NAVIN\Desktop\sample_transactional_data.csv',sep = ",")

## Checking shape of the data
print(asgn_dt.shape)
asgn_dt.head()

(268189, 12)


date  order_id  retailer_id  customer_id  medicine_id  pincode  \
0  12-03-2016    246748          119       228198       290065   400006   
1  30-01-2016    200028          520       248458       130566   400062   
2  08-01-2016    175367          252       173672       231276   110040   
3  15-03-2016    249718          612       249430       199915   110075   
4  15-03-2016    249934          469       160771       155561   400092   

   distributor_id is_chronic_flag  qtyperpack  quantity        disease_name  \
0               1           False          15        96       Multivitamins   
1               1            True          15        15  Antihyperlipidemic   
2               2            True          14        15        Antidiabetic   
3               2            True           1        16                 Ocd   
4               1           False          10        77       Anticoagulant   

  Unnamed: 11  
0         NaN  
1         NaN  
2         NaN  
3         NaN  
4         NaN

In [3]:
## Dataset info
print("Orginal Variables: ",asgn_dt.info())

## Converting numeric variables to factors and date variable to datetime object
## Date to Datetime object
asgn_dt["date"] = pd.to_datetime(asgn_dt["date"])

## ID variables to object
cols = ["order_id","retailer_id","customer_id","medicine_id","pincode","distributor_id"]
for i in cols:
    asgn_dt[i] = asgn_dt[i].astype("object")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 268189 entries, 0 to 268188
Data columns (total 12 columns):
 #   Column           Non-Null Count   Dtype 
---  ------           --------------   ----- 
 0   date             268189 non-null  object
 1   order_id         268189 non-null  int64 
 2   retailer_id      268189 non-null  int64 
 3   customer_id      268189 non-null  int64 
 4   medicine_id      268189 non-null  int64 
 5   pincode          268189 non-null  int64 
 6   distributor_id   268189 non-null  int64 
 7   is_chronic_flag  264979 non-null  object
 8   qtyperpack       268189 non-null  int64 
 9   quantity         268189 non-null  int64 
 10  disease_name     264979 non-null  object
 11  Unnamed: 11      885 non-null     object
dtypes: int64(8), object(4)
memory usage: 24.6+ MB
Orginal Variables:  None


In [4]:
## Summary of the dataset
asgn_dt.describe(include = "all")

date  order_id  retailer_id  customer_id  medicine_id  \
count                268189  268189.0     268189.0     268189.0     268189.0   
unique                   91   67702.0        312.0      23734.0      14751.0   
top     2016-03-29 00:00:00  249199.0        504.0     216069.0     250883.0   
freq                   4609      53.0       2284.0        999.0       2914.0   
first   2016-01-01 00:00:00       NaN          NaN          NaN          NaN   
last    2016-12-03 00:00:00       NaN          NaN          NaN          NaN   
mean                    NaN       NaN          NaN          NaN          NaN   
std                     NaN       NaN          NaN          NaN          NaN   
min                     NaN       NaN          NaN          NaN          NaN   
25%                     NaN       NaN          NaN          NaN          NaN   
50%                     NaN       NaN          NaN          NaN          NaN   
75%                     NaN       NaN          NaN          NaN          NaN   
max                     NaN       NaN          NaN          NaN          NaN   

         pincode  distributor_id is_chronic_flag     qtyperpack  \
count   268189.0        268189.0          264979  268189.000000   
unique     112.0             7.0               2            NaN   
top     400092.0             1.0            True            NaN   
freq     15856.0        219714.0          142608            NaN   
first        NaN             NaN             NaN            NaN   
last         NaN             NaN             NaN            NaN   
mean         NaN             NaN             NaN      10.012499   
std          NaN             NaN             NaN       8.188401   
min          NaN             NaN             NaN       0.000000   
25%          NaN             NaN             NaN      10.000000   
50%          NaN             NaN             NaN      10.000000   
75%          NaN             NaN             NaN      10.000000   
max          NaN             NaN             NaN    1000.000000   

             quantity   disease_name             Unnamed: 11  
count   268189.000000         264979                     885  
unique            NaN            303                      30  
top               NaN  Multivitamins   pregnancy termination  
freq              NaN          45054                     192  
first             NaN            NaN                     NaN  
last              NaN            NaN                     NaN  
mean        51.610394            NaN                     NaN  
std         43.478783            NaN                     NaN  
min          0.000000            NaN                     NaN  
25%         16.000000            NaN                     NaN  
50%         35.000000            NaN                     NaN  
75%         80.000000            NaN                     NaN  
max        225.000000            NaN                     NaN

#### From above we can observe that there are null values present in the dataset. Also we can see that qtyperpack has 0 values which means that there are no tablets in a single strip which cannot be possible.Also there are some instances where the quanity is 0 which might indicate that the order is been cancelled or quantity value is not available for that particular order_id. So we will exclude those cases for further analysis.

### Filtering observations where qtyperpack > 0  and quantity > 0

In [5]:
asgn_dt = asgn_dt[(asgn_dt['qtyperpack']>0) & (asgn_dt['quantity']>0)]
asgn_dt.shape

(268106, 12)

### Checking missing values in the dataset (in percentage)

In [6]:
## Percentage of Missing Values
round(asgn_dt.isnull().sum()/asgn_dt.shape[0]*100,2)

date                0.00
order_id            0.00
retailer_id         0.00
customer_id         0.00
medicine_id         0.00
pincode             0.00
distributor_id      0.00
is_chronic_flag     1.20
qtyperpack          0.00
quantity            0.00
disease_name        1.20
Unnamed: 11        99.67
dtype: float64

In [7]:
## Dropping column with 99.67% null values
asgn_dt.drop(columns = ["Unnamed: 11"],inplace = True)

### Unique Values in the dataset
print("No of Distributors: ",len(asgn_dt['distributor_id'].unique()))
print("No of Retailers: ",len(asgn_dt['retailer_id'].unique()))
print("No of Customers: ",len(asgn_dt['customer_id'].unique()))
print("No of Medicines: ",len(asgn_dt['medicine_id'].unique()))
print("No of Orders: ",len(asgn_dt['order_id'].unique()))
print("No of Diseases: ",len(asgn_dt['disease_name'].unique()))

No of Distributors:  7
No of Retailers:  312
No of Customers:  23734
No of Medicines:  14750
No of Orders:  67701
No of Diseases:  304


In [8]:
## Creating State variable based on pincode and merging with our original dataset
import pgeocode

### Extracting pincodes for Indian states
nomi = pgeocode.Nominatim('in')

### Storing unique pincodes into new dataframe
pin_state = pd.DataFrame(asgn_dt.loc[:,'pincode'].unique())
pin_state.rename(columns = {0:"pincode"},inplace = True)

### Converting Pincode to characters and extracting statenames
pin_state['pincode'] = pin_state['pincode'].astype('object')
pin_state['state'] = pin_state['pincode'].apply(lambda x: nomi.query_postal_code(x).state_name)
pin_state['state'].fillna(pin_state['state'].mode().iloc[0],inplace = True)

In [9]:
### Mapping state_names corresponding to pincode in orginal dataset
asgn_dt = pd.merge(asgn_dt,pin_state,how = "left",on = "pincode")
asgn_dt.head()

date order_id retailer_id customer_id medicine_id pincode  \
0 2016-12-03   246748         119      228198      290065  400006   
1 2016-01-30   200028         520      248458      130566  400062   
2 2016-08-01   175367         252      173672      231276  110040   
3 2016-03-15   249718         612      249430      199915  110075   
4 2016-03-15   249934         469      160771      155561  400092   

  distributor_id is_chronic_flag  qtyperpack  quantity        disease_name  \
0              1           False          15        96       Multivitamins   
1              1            True          15        15  Antihyperlipidemic   
2              2            True          14        15        Antidiabetic   
3              2            True           1        16                 Ocd   
4              1           False          10        77       Anticoagulant   

         state  
0  Maharashtra  
1  Maharashtra  
2        Delhi  
3        Delhi  
4  Maharashtra

### Imputing Null values

In [10]:
## Disease_name
asgn_dt['disease_name'].fillna("Others",inplace = True)

## is_chronic_flag
asgn_dt['is_chronic_flag'].fillna(asgn_dt['is_chronic_flag'].mode().iloc[0],inplace = True)

## Labelling chronic_flag variable
asgn_dt["chronic_flag_map"] = asgn_dt.is_chronic_flag.map({False:0,True:1})

In [11]:
## Sorting the dataset based on date and order_id
asgn_dt1 = pd.DataFrame(asgn_dt.sort_values(["date","order_id"]))

## Market Basket Analysis

In [12]:
### Data for market basket Analysis
basket = asgn_dt.groupby(['order_id','disease_name'])['quantity'].count().unstack().reset_index().fillna(0).set_index('order_id')

In [13]:
## Encoding Columns
def encode_units(x):
    if x >= 1:
        return 1
    else:
        return 0
    
basket_encode = basket.applymap(encode_units)

## Filtering transactions where medicines for more than one diseases are bought
basket_filter = basket_encode[(basket_encode > 0).sum(axis = 1) >= 2]
basket_filter

disease_name  Abdominal belt  Abdominal pain and ibs  Acid reflux  \
order_id                                                            
164411                     0                       0            0   
165367                     0                       0            0   
165532                     0                       0            0   
165789                     0                       0            0   
165851                     0                       0            0   
...                      ...                     ...          ...   
273029                     0                       0            0   
273030                     0                       0            0   
273040                     0                       0            0   
273068                     0                       0            0   
273070                     0                       0            0   

disease_name  Active inflammatory edema  Acute coronary syndrome  \
order_id                                                           
164411                                0                        0   
165367                                0                        0   
165532                                0                        0   
165789                                0                        0   
165851                                0                        0   
...                                 ...                      ...   
273029                                0                        0   
273030                                0                        0   
273040                                0                        0   
273068                                0                        0   
273070                                0                        0   

disease_name  Acute otitis media  Acute urinary retention  Adhd  \
order_id                                                          
164411                         0                        0     0   
165367                         0                        0     0   
165532                         0                        0     0   
165789                         0                        0     0   
165851                         0                        0     0   
...                          ...                      ...   ...   
273029                         0                        0     0   
273030                         0                        0     0   
273040                         0                        0     0   
273068                         0                        0     0   
273070                         0                        0     0   

disease_name  Alcohol deaddiction  Alcohol intoxication  ...  Topical nsaid  \
order_id                                                 ...                  
164411                          0                     0  ...              0   
165367                          0                     0  ...              0   
165532                          0                     0  ...              0   
165789                          0                     0  ...              0   
165851                          0                     0  ...              0   
...                           ...                   ...  ...            ...   
273029                          0                     0  ...              0   
273030                          0                     0  ...              0   
273040                          0                     0  ...              0   
273068                          0                     0  ...              0   
273070                          0                     0  ...              0   

disease_name  Travel sickness  Treatment of acute diarrhea  \
order_id                                                     
164411                      0                            0   
165367                      0                            0   
165532                      0                            0   
165789 

In [14]:
### Importing mlxtend for
from mlxtend.frequent_patterns import apriori

In [15]:
## Market Basket Analysis
freq_items = apriori(basket_filter,min_support = 0.03,use_colnames = True).sort_values('support',ascending = False).reset_index()
freq_items['length'] = freq_items['itemsets'].apply(lambda x: len(x))
freq_items

index   support                                           itemsets  \
0       15  0.492091                                    (Multivitamins)   
1       11  0.478574                                 (Antihypertensive)   
2        6  0.403180                                     (Antidiabetic)   
3       10  0.378647                               (Antihyperlipidemic)   
4       42  0.241133                   (Antidiabetic, Antihypertensive)   
..     ...       ...                                                ...   
99      27  0.031150                       (Anti bph, Antihypertensive)   
100     37  0.030888                (Antihyperlipidemic, Antidepresant)   
101     94  0.030545  (Derma product, Antihyperlipidemic, Multivitam...   
102     71  0.030525          (Antacid, Multivitamins, Antihistaminics)   
103     61  0.030101                          (Nsaid, Antihypertensive)   

     length  
0         1  
1         1  
2         1  
3         1  
4         2  
..      ...  
99        2  
100       2  
101       3  
102       3  
103       2  

[104 rows x 4 columns]

In [16]:
## Extracting rules list
from mlxtend.frequent_patterns import association_rules
rules_list = association_rules(freq_items,metric = "lift",min_threshold = 1).sort_values(['confidence'],ascending = False).reset_index(drop = True)
rules = rules_list[(rules_list['lift'] > 1)& (rules_list['confidence'] >= 0.5)]
rules

antecedents           consequents  \
0   (Antidiabetic, Antihyperlipidemic, Antihistami...    (Antihypertensive)   
1      (Antidiabetic, Multivitamins, Antihistaminics)    (Antihypertensive)   
2                           (Myoclonus, Antidiabetic)    (Antihypertensive)   
3   (Antihyperlipidemic, Multivitamins, Antihistam...    (Antihypertensive)   
4                     (Myoclonus, Antihyperlipidemic)    (Antihypertensive)   
..                                                ...                   ...   
78                                 (Antihypertensive)        (Antidiabetic)   
79                             (Steroid and hormones)    (Antihypertensive)   
80                 (Antidiabetic, Antihyperlipidemic)       (Multivitamins)   
81                  (Derma product, Antihypertensive)        (Antidiabetic)   
82                      (Antidiabetic, Multivitamins)  (Antihyperlipidemic)   

    antecedent support  consequent support   support  confidence      lift  \
0             0.051527            0.478574  0.040209    0.780345  1.630562   
1             0.059617            0.478574  0.045374    0.761083  1.590314   
2             0.051164            0.478574  0.038837    0.759069  1.586107   
3             0.049490            0.478574  0.037263    0.752956  1.573331   
4             0.072630            0.478574  0.054392    0.748889  1.564834   
..                 ...                 ...       ...         ...       ...   
78            0.478574            0.403180  0.241133    0.503857  1.249709   
79            0.088710            0.478574  0.044688    0.503753  1.052612   
80            0.195214            0.492091  0.098334    0.503721  1.023632   
81            0.072812            0.403180  0.036598    0.502632  1.246671   
82            0.196001            0.378647  0.098334    0.501698  1.324978   

    leverage  conviction  
0   0.015549    2.373836  
1   0.016842    2.182456  
2   0.014351    2.164216  
3   0.013579    2.110657  
4   0.019633    2.076475  
..       ...         ...  
78  0.048182    1.202921  
79  0.002234    1.050738  
80  0.002270    1.023433  
81  0.007241    1.199958  
82  0.024118    1.246942  

[83 rows x 9 columns]

From above, with 78% confidence we can say that retailer receiving orders for Antidiabetic, Antihyperlipidemic and Anithistamic is 1.63 times more likely to receive orders for Antihypertensive medicines 

In [18]:
## Exporting data in csv format
asgn_dt1.to_csv(r'C:\Users\NAVIN\Desktop\sample_transactional_data.csv',index = False)
rules.to_csv(r'C:\Users\NAVIN\Desktop\product_basket.csv',index = False)